share/CMakeJS.cmake

```diff
@@ -0,0 +1,640 @@
+target_link_libraries       (node-dev INTERFACE "${CMAKE_JS_LIB}")
+
+# Node API (C) - requires NodeJS system installation headers
+# cmake-js::node-api
+add_library                 (node-api INTERFACE)
+add_library                 (cmake-js::node-api ALIAS node-api)
+target_include_directories  (node-api INTERFACE "${NODE_API_HEADERS_DIR}")
+target_sources              (node-api INTERFACE "${NODE_API_INC_FILES}")
+target_link_libraries       (node-api INTERFACE cmake-js::node-dev)
+
+# Node Addon API (C++) - requires Node API (C)
+# cmake-js::node-addon-api
+add_library                 (node-addon-api INTERFACE)
+add_library                 (cmake-js::node-addon-api ALIAS node-addon-api)
+target_include_directories  (node-addon-api INTERFACE "${NODE_ADDON_API_DIR}")
+target_sources              (node-addon-api INTERFACE "${NODE_ADDON_API_INC_FILES}")
+target_link_libraries       (node-addon-api INTERFACE cmake-js::node-api)
+
+# CMakeJS API - requires Node Addon API (C++), resolves the full Napi Addon dependency chain
+# cmake-js::cmake-js
+add_library                 (cmake-js INTERFACE)
+add_library                 (cmake-js::cmake-js ALIAS cmake-js)
+target_link_libraries       (cmake-js INTERFACE cmake-js::node-addon-api)
+target_compile_definitions  (cmake-js INTERFACE "BUILDING_NODE_EXTENSION")
+target_compile_features     (cmake-js INTERFACE cxx_nullptr) # Signal a basic C++11 feature to require C++11.
+
+# Generate definitions
+if(MSVC)
+    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>" CACHE STRING "Select the MSVC runtime library for use by compilers targeting the MSVC ABI." FORCE)
+    if(CMAKE_JS_NODELIB_DEF AND CMAKE_JS_NODELIB_TARGET)
+        execute_process(COMMAND ${CMAKE_AR} /def:${CMAKE_JS_NODELIB_DEF} /out:${CMAKE_JS_NODELIB_TARGET} ${CMAKE_STATIC_LINKER_FLAGS})
+    endif()
+endif()
+
+#[=============================================================================[
+Exposes a user-side helper function for creating a dynamic '*.node' library,
+linked to the Addon API interface.
+
+cmakejs_create_napi_addon(<name> [<sources>])
+cmakejs_create_napi_addon(<name> [ALIAS <alias>] [NAMESPACE <namespace>] [NAPI_VERSION <version>] [<sources>])
+
+(This should wrap the CMakeLists.txt-side requirements for building a Napi Addon)
+]=============================================================================]#
+function(cmakejs_create_napi_addon name)
+
+    # Avoid duplicate target names
+    if(TARGET ${name})
+        message(SEND_ERROR "'cmakejs_create_napi_addon()' given target '${name}' which is already exists. Please choose a unique name for this Addon target.")
+        return()
+    endif()
+
+    set(options)
+    set(args ALIAS NAMESPACE NAPI_VERSION)
+    set(list_args)
+    cmake_parse_arguments(ARG "${options}" "${args}" "${list_args}" "${ARGN}")
+
+    # Generate the identifier for the resource library's namespace
+    set(ns_re "[a-zA-Z_][a-zA-Z0-9_]*")
+
+    if(NOT DEFINED ARG_NAMESPACE)
```

Comment:
I don't think ```NODE_API_MODULE``` actually cares about what namespace it is in. In fact, if you were to define it in a different namespace than the Addon functions, it might even fail to build unless the module specifier carries the namespaces too:

```
NODE_API_MODULE(addon, demo::addon::Init)
```

```NAPI_CPP_CUSTOM_NAMESPACE``` is part of the Node Addon API (defined in ```napi.h```). It is demonstrated in the source file attached. Implementing it is entirely up to the user (I personally never knew about it 'til recently). 

One of these bug-bears about namespaces, is that your code and your CMake targets *should* have consistent namespaces. You might be compiling a class found in ```vendor::library::class```, in which case the CMake target should also be ```vendor::library::class``` (usually done as an ALIAS  to the ```class``` target).

I mostly tested *within* the defined namespaces and getting it right was tricky :)

Will try a test project with no namespace in the source code and see if I can break it with the NAMESPACE arg, and also vice-versa.